Add in implementation for OOM retry

revans2 opened the pull request:
This is an initial implementation of the OOM retry framework. There is still a need for a lot more testing, but this is good enough to get started.

revans2 commented:
build

jbrennan333 commented on src/main/cpp/src/SparkResourceAdaptorJni.cpp:
```diff
@@ -13,37 +13,210 @@
+  }
+
+  bool operator<(const thread_priority &other) const {
+    long task_priority = std::numeric_limits<long>::max() - task_id + 1;
```
I think you need parens here:  max() - (task_id + 1)
Same in following lines.

abellina replied:
nit, may be nice to put this formula in a function so that we call that to calculate the priority.

jlowe commented on src/main/cpp/src/SparkResourceAdaptorJni.cpp:
```diff
@@ -52,121 +225,832 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
+        // fall through
+        case SHUFFLE_BLOCKED:
+          log_status("WAITING", thread_id, thread->second.task_id, thread->second.state);
+          task_wake_condition.wait(lock);
+          task_has_woken_condition.notify_all();
```
This doesn't seem to account for spurious wakeups.  Seems we need to check the state before notifying others that this thread is waking up.

jlowe commented on src/main/cpp/src/SparkResourceAdaptorJni.cpp:
```diff
@@ -52,121 +225,832 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
+        case TASK_BUFN_WAIT:
+          transition(thread->second, thread_state::TASK_BUFN);
+          log_status("WAITING", thread_id, thread->second.task_id, thread->second.state);
+          task_wake_condition.wait(lock);
+          task_has_woken_condition.notify_all();
```
Same spurious wakeup comment here.

jlowe commented on src/main/cpp/src/SparkResourceAdaptorJni.cpp:
```diff
@@ -52,121 +225,832 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
+      switch(thread->second.state) {
+        case TASK_BLOCKED:
+          transition(thread->second, thread_state::TASK_RUNNING);
+          task_wake_condition.notify_all();
```
It's a bit odd that we're notifying multiple times on the same condition variable, once per task in a blocked state.  Given the way the condition variable is being used and that I think we're always doing a state transition on the task before waking it up, seems like we could avoid a lot of unnecessary wakeups by having a condition variable per task rather than a global one.

abellina commented on src/main/cpp/src/SparkResourceAdaptorJni.cpp:
```diff
@@ -52,121 +225,832 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
+    // is effectivley blocked or not.  We could change the definitions here,
+    // but for now this sounds like a good starting point.
+    std::set<long> tasks_with_threads;
+    std::set<long> tasks_with_threads_effectively_blocked;
```
should this just be a counter? Judging by how it's used for the deadlock busting logic.

revans2 replied:
This is tasks, not threads. So if we have multiple threads on a single task a counter will not work.

abellina commented on src/main/cpp/src/SparkResourceAdaptorJni.cpp:
```diff
@@ -52,121 +225,832 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
+      for (auto thread = threads.begin(); thread != threads.end(); thread++) {
+        switch (thread->second.state) {
+          case TASK_BLOCKED:
+            {
```
style nit, shouldn't really need this top level curly bracket.

revans2 replied:
I need the top level curly to prevent `thread_priority current` from escaping that scope and causing issues in other places. I get compile errors without it.

abellina replied:
Looks like switch/case doesn't affect the scope, so that makes sense!

abellina commented on src/main/cpp/src/SparkResourceAdaptorJni.cpp:
```diff
@@ -52,121 +225,832 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
+        throw_java_exception(SPLIT_AND_RETRY_OOM_CLASS, "injected SplitAndRetryOOM");
+      }
+
+      block_thread_until_ready(lock);
```
nit, we could pass `thread_id` here... save us a lookup and check against `threads.end`  `block_thread_until_ready`

revans2 commented on src/main/cpp/src/SparkResourceAdaptorJni.cpp:
```diff
@@ -52,121 +225,832 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
   void do_deallocate(void *p, std::size_t size, rmm::cuda_stream_view stream) override {
     resource->deallocate(p, size, stream);
+    // deallocate success
+    if (size > 0) {
```
I am going to decline this right now. do_deallocate is not big enough for me to fell like I need to break it up.

revans2 commented:
I think I have addressed all of the comments. Please take another look.  Sorry. To fix the formatting question I ran the auto-format tool from CUDF and it kind of changed a lot.

revans2 commented:
build

revans2 commented:
@abellina @jlowe just pushed a fix for two possible live lock situations. One that is for a OOM when running from a non task thread and another for a CUDA fatal error that might cause all allocations to fail which would then try to retry the alloc and always loop. Please take another look

pxLi commented:
build
